### PR TITLE
Polymorphic resources

### LINKED ISSUE #754: Creating Time Series Insights environment throws InvalidRequestBody
Tried to create a Timeseriesinsights Environment

```
new azure_native.timeseriesinsights.Environment("tsi", {
    resourceGroupName: "rg1",
    kind: timeseriesinsights.EnvironmentKind.Gen2,
    sku: {
        name: timeseriesinsights.SkuName.L1,
        capacity: 1
    }

});
```

## Current behavior
Throws error on creation.

> azure-native:timeseriesinsights:Environment (tsi):
    error: Code="InvalidRequestBody" Message="Value cannot be null.\r\nParameter name: properties\r\nPath ''."

## Affected Version

azure-native@0.9.0 and lower

### LINKED ISSUE #766: blueprints/Artifact is not picking up the polymorphism
The blueprint package on the provider for its `Artifact` entity  is not properly processing the polymorphisms defined in the azure api spec. This results in one of the more releavnt entities on the package not having its required properties not being exposed in the pulumi sdks.

An example of this can be seen when looking at the artifact resource which only exposes 

```yaml
blueprintName: string
kind: string | ArtifactKind
resourceScope: string
artifactName: string
```

which comprise the definition of the base object without the polymorphism applied. Once the polimorphism is applied one would be able to define the additional parameters required to actually create a blueprint artifact for any of the `ArtifactKind`s. See [the related spec](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/blueprint/resource-manager/Microsoft.Blueprint/preview/2018-11-01-preview/blueprintDefinition.json#L757).

## Expected behavior
One is able to define the properties for each kind of artifact from within pulumi.

## Current behavior
An artifact cannot have its required properties specified within pulumi.

## Affected feature
blueprints

### PR DESCRIPTION
**Pull request by mikhailshilkov**
Resolve #766
Resolve #754

It turns out that 21 Azure resource have a discriminator property at the top level. Then, each resource has 1 to N schemas that derive from that base type and define a fixed discriminator value for the discriminated property. Up until now, we erroneously ignore those derived types, which limit the usability of the resources (often, to not usable at all).

Pulumi schema does not support `OneOf` annotation for resources: it [requires](https://github.com/pulumi/pulumi/blob/master/pkg/codegen/schema/schema.go#L740) a list of input and output properties. I think this case is highly specialized for Azure, so it's best to solve it in the provider itself.

This PR deprecates those existing resources and introduces 1 to N new resources for each of them. Each resource corresponds to a single derived type and a single discriminator value. This way, discrimination happens when a user picks a resource type, and we can fix the list of inputs and outputs for each variant.

It's a largish change, so please review carefully. Thank you!

### CODE REVIEW

File: examples/timeseries/index.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+// Copyright 2021, Pulumi Corporation.  All rights reserved.
+
+import * as resources from "@pulumi/azure-native/resources";
+import * as timeseries from "@pulumi/azure-native/timeseriesinsights";
+
+const resourceGroup = new resources.ResourceGroup("rg");
+
+new timeseries.Gen1Environment("environment", {
+    resourceGroupName: resourceGroup.name,
+    kind: "Gen1",
```
Comment (mikhailshilkov) [inline examples/timeseries/index.ts:10]: Note that users still need to specify the discriminator value here. It's the same `const` property in the schema that I use here, and that's how it currently behaves. We can track improvement for that together with https://github.com/pulumi/pulumi-azure-native/issues/647

File: provider/pkg/gen/schema.go [hunk 2]
```diff
@@ -759,6 +880,60 @@ func (g *packageGenerator) getAsyncStyle(op *spec.Operation) string {
...
+		return resolvedSchema, nil
+	}
+
+	return nil, nil
+}
+
+func getResponseSchema(ctx *openapi.ReferenceContext, statusCodeResponses map[int]spec.Response) (*openapi.Schema, error) {
+	var codes []int
+	for code := range statusCodeResponses {
+		if code >= 300 {
```
Comment (lblackstone) [inline provider/pkg/gen/schema.go:907]: Would it make sense to also filter out 1xx responses?

Comment (mikhailshilkov) [reply provider/pkg/gen/schema.go]: Done

File: CHANGELOG.md [hunk 3]
```diff
@@ -3,6 +3,178 @@ CHANGELOG
 
 ## HEAD (Unreleased)
 
+- `blueprint.PolicyAssignmentArtifact`
```
[the remark this answers is not in the document]
Comment (mikhailshilkov) [reply CHANGELOG.md:8]: Fixed

File: sdk/nodejs/blueprint/artifact.ts [hunk 4]
```diff
@@ -19,6 +21,7 @@ export class Artifact extends pulumi.CustomResource {
      * @param opts Optional settings to control the behavior of the CustomResource.
      */
     public static get(name: string, id: pulumi.Input<pulumi.ID>, opts?: pulumi.CustomResourceOptions): Artifact {
+        pulumi.log.warn("Artifact is deprecated: Artifact is deprecated. Please use one of the variants: PolicyAssignmentArtifact, RoleAssignmentArtifact, TemplateArtifact.")
```
[the remark this answers is not in the document]
Comment (mikhailshilkov) [reply sdk/nodejs/blueprint/artifact.ts]: Good catch. Removed the second instance of it.

File: sdk/nodejs/blueprint/policyAssignmentArtifact.ts [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        if (obj === undefined || obj === null) {
+            return false;
+        }
+        return obj['__pulumiType'] === PolicyAssignmentArtifact.__pulumiType;
+    }
+
+    /**
+     * Artifacts which need to be deployed before the specified artifact.
+     */
+    public readonly dependsOn!: pulumi.Output<string[] | undefined>;
```
[the remark this answers is not in the document]
Comment (mikhailshilkov) [reply sdk/nodejs/blueprint/policyAssignmentArtifact.ts:42]: I updated to the latest pulumi/pulumi but it's still there... I don't have any other context for this. Probably, not quite related to this PR?

**Pull request merged**